.github/pull_request_template.md

```diff
@@ -1,10 +1,21 @@
-Hi! Thanks for contributing to JupyterLab.
-Please look through the checklist below before submitting your pull request.
-
-- [ ] Look through open pull requests to see if someone has already started working on this.
-- [ ] Note what issue(s) this pull request addresses. (We strongly encourage the community to open an issue describing proposed work before opening a pull request.)
-- [ ] Provide a detailed description of the pull request. (Why are these changes necessary? How have they been implemented?)
-- [ ] For UI/UX changes, include before/after screenshots
-- [ ] Include a narrative description of visual or user interaction changes. How does your design effectively address the problem?
-- [ ] Provide a list of Jupyterlab packages the pull request modifies.
-- [ ] Describe any backwards incompatible changes to Jupyterlab’s public APIs.
+<!-- Thanks for contributing to JupyterLab! Please fill out the following items to submit a pull request. -->
+
+## References
+
+<!-- Note here issue numbers this pull request addresses. -->
```

Comment:
Regarding "We strongly encourage the community to open an issue … before creating a PR", this is something Brian wanted to include so contributors in the future would be more aware of and adopt this best practice of opening an issue prior to creating a PR. This ensures PRs have more feedback and are focused and relevant to the Jupyterlab community.

Reply:
How about we document these best practices in the CONTRIBUTING guide and put a link in this template? By the time they are looking here, they've already made their code changes and they are ready to submit. Creating an issue at this point is pointless, and it feels the wrong time to instruct them. Instead, if we notice it happening, we can gently direct them to the CONTRIBUTING guide for next time.

CC @ellisonbg

Reply:
Creating an issue is not pointless. Rather, if they already made the code without having had an issue, they should at the very least be made aware that they are breaking convention, but are still free to choose to still submit as is. Also, having a standard text warning you is likely better on the ego than being reprimanded after the fact for "doing it wrong".

Reply:
I'm realizing that I also disagree with the underlying convention. I think for most first-time contributors, the PR would be a small bugfix, typo in the documentation, etc., and for those small contributions I encourage people to just open a PR and not bother to open an issue. I agree that for larger contributions, especially ones that take time to implement, require design decisions, etc., opening an issue first is encouraged.

However, to move forward, how about we change this to: 

```
<!-- Note here issue numbers this pull request addresses (should be at least one, see contributing guidelines noted above). -->
```

Reply:
> How about we document these best practices in the CONTRIBUTING guide and put a link in this template?

I just did this, and noted that there should be at least one issue referenced. Does that look good to everyone?